### DISCUSSION TITLE
unexpected behavior from regexp substitution

### DISCUSSION THREAD

**Discussion by JeffMelton** [lmorg/murex#905] [Q&A] [answered]
I haven't seen this behavior documented anywhere, so I suspect it's a bug, but I'll start here in case it's PEBKAC. I wanted to remove all leading zeroes from a string of integers, so `out "0123" -> regexp 's/^0*//'` seemed sensible. Documentation for `regexp` indicates that the `^` anchor for beginning of line is supported, but `^0` seems to be some kind of syntactic sugar (undocumented, perhaps?) for "very first history entry". 

[![asciicast](https://asciinema.org/a/DfW4BOMAfqO42yXIMxPQ7sETC.svg)](https://asciinema.org/a/DfW4BOMAfqO42yXIMxPQ7sETC)

I was able to get what I needed with `'s/\A0*//'`, but I figured this was worth asking about.

**Comment by lmorg** [comment, marked as answer]
You're right on both counts.

`^` is used as a token for history variables. And you're right that it's undocumented.

It's also amongst some of Murex's oldest code: https://github.com/lmorg/murex/blob/master/shell/history/vars.go

Given how old the code is and how it's undocumented, I'm definitely open to revisiting / reworking this implementation.

One thought I'd been having lately was having this a feature of `readline`. Making it a command mode like `vi`'s `:` command mode. So it could be invoked via `[esc]:`

**Reply by JeffMelton** [reply]
I don't have strong feelings about it one way or the other, I suppose. I don't have much use for that `history` feature myself, but I do lean heavily on regex anchors, and I hadn't been aware of the `\A` one until now. `^` and `$` are carved into my muscle memory right next to Vi motion keybindings. 😄 

I would say that until you've decided what you want to do with that feature the docs should probably be updated in the `regexp` section to point out the quirks around using `^`, since that applies not only to the beginning-of-line anchor but also to character class negation. Is there a fancy quoting or escape approach that would avoid triggering that history feature, or is there a config flag for disabling it? I tried a single backslash, but once I found `\A` I quit playing with it.

**Reply by lmorg** [reply]
Presently there's no way to escape nor disable it. The only saving grace is that it only applies to stuff typed in the REPL. `read` doesn't use it and nor do shell scripts.

To be honest, I never completely liked the `^` pattern. It made sense back when Murex was in beta and I used a lot more "off-the-shelf" libraries. But now I've rewritten everything else, it's time for me to rethink this too.

Some of the functions, like `^s/find/replace/` definitely don't belong as prompt variables and should be integrated into `readline` instead.

**Reply by JeffMelton** [reply]
Ah, ok, thanks. 🙏🏽

**Reply by lmorg** [reply]
The history variables haven't been replaced, however a first draft of porting that functionality to `readline` is in the following branch: https://github.com/lmorg/murex/pull/907

Here, you can access the command palette via `{ESC}:` (like you would in `vim` too)

Supported commands are:
```
"Valid commands:":   "",
"  !!":              "Previous command line",
"  n":               "history item n (where 'n' is an integer)",
"  m/find/":         "match string (regexp)",
"  s/find/replace/": "substitute in current line (regexp)",
```
however more can and will follow
